python/plugins/processing/core/parameters.py

```diff
@@ -79,13 +80,25 @@ def setValue(self, obj):
         of parameter.
         """
         if obj is None:
-            if not self.optional:
+            if not self.optional and self.default is None:
                 return False
-            self.value = None
-            return True
+            elif self.default is None:
+                self.value = None
+                return True
+            else:
```

Comment:
My reading of this code is that if you have a non-None default, you can't assign None to a parameter value.

This is contrary to my understanding of our conclusion here: https://github.com/qgis/QGIS/commit/3472ac80d8bdd030090ce81cb686f3cfcbd913cd

This is the current behavior (pretend for a moment that Parameter takes an optional "default" parameter - maybe it should).
```
param = Parameter(optional=True, default="Foo")
assert(param.setValue(None))
assertEquals(param.value, "Foo")
```


Here is the behavior I would expect
```
param = Parameter(optional=True, default="foo")
assertEqual(param.value, "foo")
assert(param.setValue(None))
assertEqual(param.value, None)
```

Am I mistaken?